File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Joaquin/ArmJoint/ArmJointConfig.java
package org.firstinspires.ftc.teamcode.Joaquin.ArmJoint;

import com.qualcomm.robotcore.hardware.DcMotorSimple;
import org.firstinspires.ftc.teamcode.utils.Reduction;
import com.seattlesolvers.solverslib.controller.PController;


public class ArmJointConfig {

    private final String motorId;
    private final int maximumAngleLimit;
    private final int minimumAngleLimit;
    private final Reduction gearRatio;
    private final int tolerance;
    private final PController pController;



    public ArmJointConfig(String motorId, int maximumAngleLimit, int minimumAngleLimit, Reduction gearRatio, int tolerance, PController pController){
        this.motorId = motorId;
        this.maximumAngleLimit = maximumAngleLimit;
        this.minimumAngleLimit = minimumAngleLimit;
        this.gearRatio = gearRatio;
        this.tolerance = tolerance;
        this.pController = pController;
    }

    public String getMotorId(){
        return motorId;
    }
    public int getMaximumAngleLimit(){
        return maximumAngleLimit;
    }
    public int getMinimumAngleLimit(){
        return minimumAngleLimit;
    }
    public int getTolerance(){
        return tolerance;
    }

    public Reduction getGearRatio(){
        return gearRatio;
    }

    public PController getpController() {
        return pController;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/arm/armJoint/ArmJointConfig.kt
package org.firstinspires.ftc.ulateamcode.Arm

import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.seattlesolvers.solverslib.controller.PIDController
import org.firstinspires.ftc.teamcode.utils.Reduction
import com.seattlesolvers.solverslib.hardware.motors.Motor

data class ArmJointConfig(
    val motorId: String,
    val motorType: Motor.GoBILDA,
    val motorDirection: DcMotorSimple.Direction,
    val gearRatio: Reduction,
    val bottomAngleLimit : Double,
    val upperAngleLimit : Double,
    val pController : PIDController,
    val zeroPowerBehavior : Motor.ZeroPowerBehavior,
    val runMode: Motor.RunMode
)

/**
 * Still need to configure joint with the correct motor type, direction, reduction, bottom and superior angle
 * limits and the P controller.
 */
val bottomJointConfig = ArmJointConfig(
    motorId = "bottomLinkMotor",
    motorType = Motor.GoBILDA.RPM_1150,
    motorDirection = DcMotorSimple.Direction.FORWARD,
    gearRatio = Reduction(0.0),
    bottomAngleLimit = 0.0,
    upperAngleLimit = 0.0,
    pController = PIDController(0.0, 0.0, 0.0),
    zeroPowerBehavior = Motor.ZeroPowerBehavior.BRAKE,
    runMode = Motor.RunMode.PositionControl
)

/**
 * Still need to configure joint with the correct motor type, direction, reduction, bottom and superior angle
 * limits and the P controller.
 */
val upperJointConfig = ArmJointConfig(
    motorId = "upperLinkMotor",
    motorType = Motor.GoBILDA.RPM_1150,
    motorDirection = DcMotorSimple.Direction.REVERSE,
    gearRatio = Reduction(0.0),
    bottomAngleLimit = 0.0,
    upperAngleLimit = 0.0,
    pController = PIDController(0.0, 0.0, 0.0),
    zeroPowerBehavior = Motor.ZeroPowerBehavior.BRAKE,
    runMode = Motor.RunMode.PositionControl
)
